SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AutodiffClosureSpecialization.swift

```diff
@@ -0,0 +1,180 @@
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===-----------------------------------------------------------------------===//
+
+/// AutoDiff Closure Specialization
+/// ----------------------
+/// This optimization performs closure specialization tailored for the patterns seen
+/// in Swift Autodiff. In principle, the optimization does the same thing as the existing
+/// closure specialization pass. However, it is tailored to the patterns of Swift Autodiff.
+/// 
+/// The compiler performs reverse-mode differentiation on functions marked with `@differentiable(reverse)`.
+/// In doing so, it generates corresponding VJP and Pullback functions, which perform the 
+/// forward and reverse pass respectively. The Pullbacks are essentially a chain of closures, 
+/// where the closure-contexts are used as an implicit "tape". 
+/// 
+/// The code patterns that this optimization targets, look similar to the one below:
+/// ``` swift
+/// @differentiable(reverse)
+/// func foo(_ x: Float) -> Float { 
+///     return sin(x)
+/// }
+/// 
+/// //============== Before optimization ==============//
+/// sil @pb_foo: $(Float, (Float) -> Float) -> Float { 
+/// bb0(%0: $Float, %1: $(Float) -> Float):
+///     %2 = apply %1(%0): $Float
+///     return %2: $Float
+/// }
+/// 
+/// sil @vjp_foo: $(Float) -> (Float, (Float) -> Float) { 
+/// bb0(%0: $Float):
+///     %1 = apply @sin(%0): $Float
+///     %2 = partial_apply @pb_sin(%0): $(Float) -> Float
+///     %pb_foo = partial_apply @pb_foo(%2): $(Float, (Float) -> Float) -> Float
+///     return (%y3, %pb_foo)
+/// }
+/// 
+/// //============== After optimization ==============//
+/// sil @specialized_pb_foo: $(Float, Float) -> Float { 
+/// bb0(%0: $Float, %1: $Float):
+///     %2 = partial_apply @pb_sin(%1): $(Float) -> Float
+///     %3 = apply %2(%0): $Float
+///     return %3: $Float
+/// }
+/// 
+/// sil @vjp_foo: $(Float) -> (Float, (Float) -> Float) { 
+/// bb0(%0: $Float):
+///     %1 = apply @sin(%0): $Float
+///     %pb_foo = partial_apply @specialized_pb_foo(%0): $(Float, Float) -> Float
+///     return (%y3, %pb_foo)
+/// }
+/// ``` 
+///
+/// Similar complications, as in the existing closure-specialization optimization, exist
+/// when populating the specialized Pullback and rewriting the apply instruction in the 
+/// VJP. The details of the complications and our work arounds are listed below 
+/// (copied over from the existing closure-specialization pass):
+///
+/// 1. If we support the specialization of closures with multiple user callsites
+///    that can be specialized, we need to ensure that any captured values have
+///    their reference counts adjusted properly. This implies for every
+///    specialized call site, we insert an additional retain for each captured
```

Comment:
I think it is fine if we only support OSSA in the new pass (`copy_value` instead of `strong_retain`). It's the plan to move OSSA down the optimizer pipeline. And for autodiff  we can probably run this pass earlier where we already have OSSA.

Reply:
I actually think we're going to have to reuse the logic in existing pass for this. Since the `ClosureSpecCloner` is what handles the callee side of reference count adjustments and we'd be bridging to it because SILCloner is not available in Swift.